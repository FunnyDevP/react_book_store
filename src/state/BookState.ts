import { atom, selector } from "recoil";
import { Cart } from "../models/Cart";

export const bookState = atom<Cart[]>({
  key: "carts",
  default: [],
});

export const bookSelector = selector({
  key: "valueCartsState",
  get: ({ get }) => {
    return get(bookState);
  },
});
