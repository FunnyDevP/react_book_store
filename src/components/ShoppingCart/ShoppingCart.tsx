import React, { FC, useState } from "react";
import { Button, Dialog, Divider, Grid, Slide } from "@mui/material";
import { TransitionProps } from "@mui/material/transitions";
import Typography from "@mui/material/Typography";
import { useRecoilValue } from "recoil";
import { bookSelector } from "../../state/BookState";

const Transition = React.forwardRef(function Transition(
  props: TransitionProps & {
    children: React.ReactElement<any, any>;
  },
  ref: React.Ref<unknown>
) {
  return <Slide direction="up" ref={ref} {...props} timeout={450} />;
});

const ShoppingCart: FC = () => {
  const [isOpen, setIsOpen] = useState(false);
  const resultBook = useRecoilValue(bookSelector);

  const handleClickOpen = () => {
    setIsOpen(true);
  };
  const handleClose = () => {
    setIsOpen(false);
  };
  return (
    <>
      <Button variant="contained" onClick={handleClickOpen}>
        Cart
      </Button>
      <Dialog
        sx={{
          "& .MuiDialog-paper": {
            width: "100%",
            maxHeight: 700,
            maxWidth: 1000,
          },
        }}
        open={isOpen}
        TransitionComponent={Transition}
        keepMounted
        onClose={handleClose}
      >
        <Grid container>
          <Grid item xs={8} sx={{ padding: "36px 36px 36px 36px" }}>
            <Typography variant={"h6"}>Shopping cart</Typography>
            <Divider
              sx={{
                marginTop: "10px",
                marginBottom: "20px",
                borderColor: "black",
              }}
            />
            {resultBook.length == 0 ? (
              <Typography variant={"h5"}>
                There are no items in your cart. Find an item in Home page that
                meets your needs.
              </Typography>
            ) : (
              resultBook.map((cart) => (
                <Grid
                  item
                  xs={12}
                  data-testid={"result_book_id"}
                  key={cart.bookId}
                >
                  <Typography variant={"subtitle1"}>{cart.bookName}</Typography>
                  <Typography variant={"subtitle1"}>
                    {cart.bookPrice}
                  </Typography>
                </Grid>
              ))
            )}
          </Grid>
          {resultBook.length != 0 && (
            <Grid
              item
              xs={4}
              sx={{
                backgroundColor: "#eeeeee",
                padding: "18px 18px 18px 18px",
              }}
            >
              <Typography variant={"subtitle1"} sx={{ fontWeight: "bold" }}>
                summary {resultBook.length} items
              </Typography>
              <Grid container>
                <Grid item xs={3}>
                  <Typography variant={"h5"} sx={{ fontWeight: "bold" }}>
                    Total
                  </Typography>
                </Grid>
                <Grid item xs={9}>
                  <Typography
                    variant={"h5"}
                    sx={{ fontWeight: "bold", textAlign: "end" }}
                  >
                    {resultBook.length > 0 &&
                      resultBook.reduce((acc, cart) => acc + cart.bookPrice, 0)}
                  </Typography>
                </Grid>
              </Grid>
            </Grid>
          )}
        </Grid>
      </Dialog>
    </>
  );
};

export default ShoppingCart;
