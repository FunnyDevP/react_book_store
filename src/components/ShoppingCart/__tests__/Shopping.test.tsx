import { fireEvent, render, screen, within } from "@testing-library/react";
import ShoppingCart from "../ShoppingCart";
import { RecoilRoot, useRecoilState } from "recoil";
import { bookState } from "../../../state/BookState";
import { useEffect } from "react";

const RecoilObserver = () => {
  const expectedData = [
    {
      bookId: "405f49b6-e570-4081-8c51-861dd6bfac65",
      bookCateId: "b3d9e7a8-3c31-4aed-9984-c65c14ef0795",
      bookName: "Getting Started with Natural Language Processing",
      bookPrice: 19.99,
    },
    {
      bookId: "c1896798-5c34-4343-a5af-8b0eff2371a3",
      bookCateId: "b3ce1341-d10c-429f-954b-854f55aef90b",
      bookName: "Grokking Functional Programming",
      bookPrice: 28.79,
    },
  ];
  const [, setState] = useRecoilState(bookState);
  useEffect(() => setState(expectedData), []);
  return null;
};
describe("shopping cart component", () => {
  it("should open modal when click cart button", () => {
    const { baseElement, queryByRole } = render(
      <RecoilRoot>
        <ShoppingCart />
      </RecoilRoot>
    );

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fireEvent.click(queryByRole("button"));
    expect(baseElement).toMatchSnapshot();

    expect(screen.queryByText("Shopping cart")).toBeInTheDocument();
  });
  it("should display result book", () => {
    render(
      <RecoilRoot>
        <RecoilObserver />
        <ShoppingCart />
      </RecoilRoot>
    );

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fireEvent.click(screen.queryByRole("button"));
    expect(screen.queryByText("Shopping cart")).toBeInTheDocument();
    const resultListBook = screen.getAllByTestId("result_book_id");
    const list1 = within(resultListBook[0]);
    expect(
      list1.queryByText("Getting Started with Natural Language Processing")
    ).toBeInTheDocument();
    expect(list1.getByText("19.99")).toBeInTheDocument();
  });
  it("should count item and display", () => {
    render(
      <RecoilRoot>
        <RecoilObserver />
        <ShoppingCart />
      </RecoilRoot>
    );

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fireEvent.click(screen.queryByRole("button"));
    expect(screen.queryByText("Shopping cart")).toBeInTheDocument();
    expect(screen.queryByText("summary 2 items")).toBeInTheDocument();
  });
  it("should summary book price and display", () => {
    render(
      <RecoilRoot>
        <RecoilObserver />
        <ShoppingCart />
      </RecoilRoot>
    );

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fireEvent.click(screen.queryByRole("button"));
    expect(screen.queryByText("Shopping cart")).toBeInTheDocument();
    expect(screen.queryByText("48.78")).toBeInTheDocument();
  });
  it("should render message no items... when shopping cart is empty", () => {
    render(
      <RecoilRoot>
        <ShoppingCart />
      </RecoilRoot>
    );

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    fireEvent.click(screen.queryByRole("button"));
    expect(screen.queryByText("Shopping cart")).toBeInTheDocument();
    expect(
      screen.queryByText(
        "There are no items in your cart. Find an item in Home page that meets your needs."
      )
    ).toBeInTheDocument();
  });
});
