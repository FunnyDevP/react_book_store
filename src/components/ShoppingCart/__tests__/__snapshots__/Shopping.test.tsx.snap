// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shopping cart component should open modal when click cart button 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Cart
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiModal-root MuiDialog-root css-6asfur-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 450ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 450ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="mui-428698438"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-1gesvnc-MuiGrid-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Shopping cart
            </h6>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-8l4389-MuiDivider-root"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-ul9tu0-MuiGrid-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-1isisa7-MuiTypography-root"
            >
              summary 
              0
               items
            </h6>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1fdpgu7-MuiGrid-root"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-bezmlo-MuiTypography-root"
                >
                  Total
                </h5>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9 css-l4g9cp-MuiGrid-root"
              />
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-a0vo9r-MuiGrid-root"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-bezmlo-MuiTypography-root"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
